### ISSUE TITLE
coc-rust-analyzer doesn't display "related information" diagnostics ("value moved here" etc)

### ISSUE THREAD

**Issue by Ferdi265** [fannheyward/coc-rust-analyzer#174]
`coc-rust-analyzer` doesn't display the "related information" part of diagnostic messages, even though it is reported by the `rust-analyzer` language server:

Example program: ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=50b36351163e927701c48976cf65a876))
```rust
fn main() {
    let x = "hello world!\n".to_string();
    //  - move occurs because `x` has type `std::string::String`, which does not implement the `Copy` trait
    // (this diagnostic is missing)

    take(x);
    //   - value moved here
    // (this diagnostic is missing)

    println!("{}", x);
    //             ^ value borrowed here after move
    // (this diagnostic is displayed)
}

fn take(s: String) {
    println!("{}", s);
}
```

Diagnostic information from trace:
```
[Trace - 18:21:52] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "diagnostics": [
        {
            "code": "E0382",
            "message": "borrow of moved value: `x`\nvalue borrowed here after move",
            "range": {
                "end": {
                    "character": 20,
                    "line": 5
                },
                "start": {
                    "character": 19,
                    "line": 5
                }
            },
            "relatedInformation": [
                {
                    "location": {
                        "range": {
                            "end": {
                                "character": 10,
                                "line": 3
                            },
                            "start": {
                                "character": 9,
                                "line": 3
                            }
                        },
                        "uri": "file:///home/yrlf/Programming/rs/misc/foo/src/main.rs"
                    },
                    "message": "value moved here"
                },
                {
                    "location": {
                        "range": {
                            "end": {
                                "character": 9,
                                "line": 1
                            },
                            "start": {
                                "character": 8,
                                "line": 1
                            }
                        },
                        "uri": "file:///home/yrlf/Programming/rs/misc/foo/src/main.rs"
                    },
                    "message": "move occurs because `x` has type `std::string::String`, which does not implement the `Copy` trait"
                }
            ],
            "severity": 1,
            "source": "rustc"
        }
    ],
    "uri": "file:///home/yrlf/Programming/rs/misc/foo/src/main.rs"
}
```

Current display in CoC with `coc-rust-analyzer`:
![view in editor](https://i.imgur.com/dfDOPZz.png)

**Comment by fannheyward** [comment]
coc.nvim + coc-rust-analyzer already support `relatedInformation` diagnostic, but not displayed in floating window. Put your cursor on line 6 `x`, current diagnostic position, run `:CocCommand workspace.diagnosticRelated`, you will get related informations in coc lists.

I'll try to improve this.

**Comment by fannheyward** [comment]
Try `:CocUpdate`. I've displayed `relatedInformation` messages with diagnostic message. You still need `:CocCommand workspace.diagnosticRelated` to do jumping.

**Comment by Ferdi265** [comment]
Thanks! Much more usable now :)

I usually use the virtual text feature more than the floating messages, so seeing them there would be the absolute best, but I think I'll try experimenting with the plugin code myself and report my findings when I did :)

**Comment by Ferdi265** [comment]
With a bit of experimentation I managed to create separate diagnostics from the related information.

![virtual text related information](https://user-images.githubusercontent.com/4077106/77646071-1ea89980-6f64-11ea-9129-d14e249f7bd3.png)

IMO this should probably be solved directly in [neoclide/coc.nvim](https://github.com/neoclide/coc.nvim) though, since many language servers publish related information, which might be useful to see in-editor.

```patch
diff --git a/src/client.ts b/src/client.ts
index 4f749af..fdc6862 100644
--- a/src/client.ts
+++ b/src/client.ts
@@ -40,18 +40,26 @@ export function createClient(config: Config, bin: string): LanguageClient {
         return help;
       },
       handleDiagnostics: async (uri, diagnostics, next) => {
+        const mkInfoDiag = (related, diag) => {
+          return {
+            "code": diag.code,
+            "source": diag.source,
+            "message": related.message,
+            "range": related.location.range,
+            "uri": related.location.uri,
+            "severity": 4
+          };
+        };
+        const infos = [];
+
         diagnostics.map((diagnostic) => {
           if (diagnostic.relatedInformation) {
-            let message = `${diagnostic.message}\n\nRelated diagnostics: (Run \`:CocCommand workspace.diagnosticRelated\` to jump)\n`;
             for (const info of diagnostic.relatedInformation) {
-              const basename = path.basename(Uri.parse(info.location.uri).fsPath);
-              const ln = info.location.range.start.line;
-              message = `${message}\n${basename}(line ${ln + 1}): ${info.message}`;
+              infos.push(mkInfoDiag(info, diagnostic.code));
             }
-            diagnostic.message = message;
           }
         });
-        next(uri, diagnostics);
+        next(uri, diagnostics.concat(infos));
       },
     },
     outputChannel,
```

**Comment by fannheyward** [comment]
I was going to concat related information as diagnostics to all diagnostics list, same as what your patch does, but this is how VSCode + rust-analyzer does:

<img width="594" alt="Screen Shot 2020-03-26 at 20 54 30" src="https://user-images.githubusercontent.com/345274/77648988-ff316100-6fa3-11ea-8812-f8205fbde791.png">

I don't know whether it's correct to separate the related informations to diagnostics list.

**Comment by Ferdi265** [comment]
At least for me, fixing bugs in a larger function is much easier when the related diagnostics are also shown directly in the editor.

It would be cool to have it as an option, so users can choose whether or not they should be shown as separate diagnostics.

Also, integrating the related information into the popup should IMO also be solved in [coc.nvim](https://github.com/neoclide/coc.nvim), since it might be useful to have the message structured (this would allow switching between both styles for any language via a config option).

**Comment by fannheyward** [comment]
Agreed, showing related informations will do grep help for user. We should not be limited to how VSCode does.

**Comment by fannheyward** [comment]
`diagnostic.separateRelatedInformationAsDiagnostics` has added into coc.nvim.
